pkg/endpoints/router.go

```diff
@@ -98,6 +113,72 @@ func (r Resource) RegisterReadinessProbes(container *restful.Container) {
 	container.Add(wsv4)
 }
 
+// Back-end extension: Requests to the URL are passed through to the Port of the Name service (extension)
+// "label: tekton-dashboard-extension=true" in the service defines the extension
+// "annotation: tekton-dashboard-endpoints=<URL>" specifies the path for the extension
+type Extension struct {
+	Name           string  `json:"name"`
+	URL            string  `json:"url"`
+	Port           string  `json:"port"`
+	DisplayName    string  `json:"displayname"`
+	BundleLocation string  `json:"bundlelocation"`
+}
+var	extensions     []Extension
+
+// RegisterExtension - this discovers the extensions and registers them as the REST API extension 
+func (r Resource) RegisterExtension(container *restful.Container, namespace string) {
+	logging.Log.Info("Adding API for extensions")
+	svcs, err := r.K8sClient.CoreV1().Services(namespace).List(metav1.ListOptions{LabelSelector: extensionLabel})
+	if err != nil {
+		logging.Log.Errorf("no extension found: %s", err)
+		return
+	}
+	ws := new(restful.WebService)
+	ws.
+		Path("/").
+		Consumes(restful.MIME_JSON).
+		Produces(restful.MIME_JSON)
+
+	for _, svc := range svcs.Items {
+		url, ok := svc.ObjectMeta.Annotations[urlKey]
+		if ok {
+			logging.Log.Debugf("extension URL: %s", url)
+			ext := Extension { Name: svc.ObjectMeta.Name, URL: url, Port: getPort(svc),
+ 				DisplayName: svc.ObjectMeta.Annotations[displayNameKey],
+				BundleLocation: svc.ObjectMeta.Annotations[bundleLocationKey]}
+			// extension handler is registered at the url
+			ws.Route(ws.GET(url).To(ext.HandleExtension))
+			ws.Route(ws.POST(url).To(ext.HandleExtension))
+			ws.Route(ws.PUT(url).To(ext.HandleExtension))
+			ws.Route(ws.DELETE(url).To(ext.HandleExtension))
+			extensions = append(extensions, ext)
+		}
+	}
+	logging.Log.Debugf("extension: %+v", extensions)
+	container.Add(ws)
+}
+
+// HandleExtension - this routes request to the extension service
+func (ext Extension) HandleExtension(request *restful.Request, response *restful.Response) {
+	target, _ := url.Parse("http://" +  ext.Name + ":" + ext.Port + "/")
+	proxy := httputil.NewSingleHostReverseProxy(target)
+	proxy.ServeHTTP(response, request.Request)
+}
+
+/* Get all extensions in a given namespace */
+func (r Resource) getAllExtensions(request *restful.Request, response *restful.Response) {
+	namespace := request.PathParameter("namespace")
+	logging.Log.Debugf("In getAllExtensions: namespace: %s", namespace)
+	logging.Log.Debugf("extension: %+v", extensions)
+
```

Comment:
General note for all changes ..... inconsistency in messages.

Debug/Info should probably start with a capital - error messages should be all lower case

Reply:
Is it at all possible that the UI might be trying to use this REST endpoint to read the extensions while extensions are still being processed by main.go? .... if so protection required - don't think this is currently an issue as we don't serve content from main.go until after processing extensions has completed, but if this ever changed to be an informer rather than a static listing protection would be needed.